code/Database.md

```diff
@@ -0,0 +1,114 @@
+# Database
+
+Well database is a database. It is responsible for providing persistence and ensuring data-consistency.
+
+## Master Database
+
+Every peer needs a collection of databases to work. This collection is called an master database.
+
+It links together all the sub-system specific databases. It places no specific requirement how the databases are implemented. It can be one or multiple backends, whichever is more useful for a particular database.
+
+```
+// DB is the master database for the satellite
+type DB interface {
+	// CreateTables initializes the database
+	CreateTables() error
+	// Close closes the database
+	Close() error
+
+	// Overlay returns database for caching overlay information
+	Overlay() overlay.DB
+}
+```
+
+The interface definition is located in the specific peer folder `<peer>/peer.go`. As an example peer `satellite` has a master database with fully qualified name `satellite.DB`.
+
+## Interface
+
+Every sub-system declares an interface what it needs from persistence system. This is defined close to the sub-system using the database.
+
+For example this is a partial declaration of overlay cache database:
+
+```
+// DB is for storing node information.
+type DB interface {
+	// Get looks up the node by nodeID
+	Get(ctx context.Context, nodeID storj.NodeID) (*pb.Node, error)
+
+    // SelectStorageNodes looks up nodes based on criteria
+	SelectStorageNodes(ctx context.Context, count int, criteria *NodeCriteria) ([]*pb.Node, error)
+}
+```
+
+_Database interfaces may implement domain specific concerns, if there's a significant performance benefit._
+
+## Implementation
+
+Each master database has an implementation. The master database and sub-databases are located in `<peer>/<peer>db`. As an example implementation for `satellite.DB` interface is at `satellite/satellitedb/*.go`.
+
+Often the databases have in-memory variants, which are useful for testing. They can also support different backends, such as SQLite and Postgres. For satellite the function `satellite/satellitedb/satellitedbtest.Run` loops through different backends making testing easier. Storage Node has a similar func.
+
+Master Database will return a specific struct to fulfill the specific database interfaces.
+
+## dbx code generation
+
+For satellite we also use [`dbx`](https://github.com/spacemonkeygo/dbx/) which helps to support Sqlite and Postgres at the same time. Of course, this complicates some of the queries as a result.
```

Comment:
for consistency, it might be nice to always use `SQLite` or `Sqlite`, but not a mixture of both